autovacuum_message_attachment/models/autovacuum_mixin.py

```diff
@@ -0,0 +1,71 @@
+# Copyright (C) 2019 Akretion
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+import logging
+
+import odoo
+from odoo import api, models
+from odoo.tools.safe_eval import safe_eval
+import datetime
+
+_logger = logging.getLogger(__name__)
+
+
+class AutovacuumMixin(models.AbstractModel):
+    _name = "autovacuum.mixin"
+    _description = "Mixin used to delete messages or attachments"
+
+    @api.multi
+    def batch_unlink(self):
+        with api.Environment.manage():
+            with odoo.registry(
+                    self.env.cr.dbname).cursor() as new_cr:
+                new_env = api.Environment(new_cr, self.env.uid,
+                                          self.env.context)
+                try:
+                    while self:
+                        batch_delete = self[0:1000]
+                        self -= batch_delete
+                        # do not attach new env to self because it may be
+                        # huge, and the cache is cleaned after each unlink
+                        # so we do not want to much record is the env in
+                        # which we call unlink because odoo would prefetch
+                        # fields, cleared right after.
+                        batch_delete.with_env(new_env).unlink()
+                        new_env.cr.commit()
+                except Exception as e:
+                    _logger.exception(
+                        "Failed to delete Ms : %s" % (self._name, str(e)))
+
+    # Call by cron
+    @api.model
+    def autovacuum(self, ttype='message'):
+        rules = self.env['vacuum.rule'].search([('ttype', '=', ttype)])
+        for rule in rules:
+            records = rule._search_autovacuum_records()
+            records.batch_unlink()
+
+    def _get_autovacuum_domain(self, rule):
+        return []
+
+    def _get_autovacuum_records(self, rule):
+        if rule.model_id and rule.model_filter_domain:
+            return self._get_autovacuum_records_model(rule)
+        return self.search(self._get_autovacuum_domain(rule))
+
+    def _get_autovacuum_records_model(self, rule):
+        domain = self._get_autovacuum_domain(rule)
+        record_domain = safe_eval(rule.model_filter_domain,
+                                  locals_dict={'datetime': datetime})
+        autovacuum_relation = self._autovacuum_relation
+        for leaf in domain:
+            if not isinstance(leaf, (tuple, list)):
+                record_domain.append(leaf)
+                continue
+            field, operator, value = leaf
+            record_domain.append(
+                ('%s.%s' % (autovacuum_relation, field), operator, value))
+        records = self.env[rule.model_id.model].search(record_domain)
+        return self.search(
+            domain + [('res_id', 'in', records.ids)]
```

Comment:
I think we have some issues here, because I guess (did not test) this code only work when a rule is link to 1 specific model.
An it is not always the case. The rule coud concerns multiple models, or none. Then, we don't have `model_filter_domain`.
Then `record_domain` is empty., so I guess Odoo will fail in these case.

Anyway I think we have to avoid this logic if we don't have any `model_filter_domain`.
We need to refactore this part.
Something like 
   def _get_autovacuum_records_model(self, rule):
        domain = self._get_autovacuum_domain(rule)
        if rule.model_id and rule.model_filter_domain:
            record_domain = safe_eval(rule.model_filter_domain,
                                  locals_dict={'datetime': datetime})
            autovacuum_relation = self._autovacuum_relation
            for leaf in domain:
                if not isinstance(leaf, (tuple, list)):
                    record_domain.append(leaf)
                    continue
                field, operator, value = leaf
                record_domain.append(
                    ('%s.%s' % (autovacuum_relation, field), operator, value))
            records = self.env[rule.model_id.model].search(record_domain)
            domain +=  [('res_id', 'in', records.ids)]
        return self.search(
            domain)

Reply:
_get_autovacuum_records_model is only called if we have a domain and one model, You can check on `_get_autovacuum_records`.

I made the changes locally and the tests from other modules are executed properly. I will test on travis.

Reply:
Oh right, should be good then

Reply:
https://travis-ci.org/tegin/server-tools/builds/640488894?utm_medium=notification&utm_source=email

Here you can check that with only my change travis is fixed :wink:

Reply:
Great, can you add the string to the `model_id` field and make a pull request?
Thanks for the work

Reply:
akretion#23